### PR TITLE
Client Websockets

### PR DESCRIPTION
**Pull request by basicer**
## Related Ticket(s)
- Works toward #3537

## Short roundup of the initial problem
- Client can't connect over websockets

## What will change with this Pull Request?
- When selecting a websocket port (80, 443, 8080, 4748) . connect using webosckets instead of raw TCP.
- Servatrice can read the client IP out of an HTTP header to support reverse proxies in front of it.

### CODE REVIEW

**Comment by ZeldaZach** [comment]
screenshot 2019-02-01 15 18 23

Could be a problem

File: cockatrice/src/remoteclient.cpp [hunk 1]
```diff
@@ -301,21 +319,51 @@ void RemoteClient::readData()
...
+        buf.data()[1] = (unsigned char)(size >> 16);
+        buf.data()[0] = (unsigned char)(size >> 24);
+
+        socket->write(buf);
+    }
+}
+
+void RemoteClient::connectToHost(const QString &hostname, unsigned int port)
+{
+    usingWebSocket = port == 443 || port == 80 || port == 4748 || port == 8080;
```
Comment (ctrlaltca) [inline cockatrice/src/remoteclient.cpp]: maybe it's easier to think that anything but 4747 is using websockets?
`usingWebSocket = port != 4747;`

File: servatrice/src/serversocketinterface.cpp [hunk 2]
```diff
@@ -1037,14 +1037,15 @@ Response::ResponseCode AbstractServerSocketInterface::cmdRegisterAccount(const C
         return Response::RespUserAlreadyExists;
     }
 
-    if (servatrice->getMaxAccountsPerEmail() &&
-        !(sqlInterface->checkNumberOfUserAccounts(emailAddress) < servatrice->getMaxAccountsPerEmail())) {
-        if (servatrice->getEnableRegistrationAudit())
-            sqlInterface->addAuditRecord(QString::fromStdString(cmd.user_name()).simplified(), this->getAddress(),
-                                         QString::fromStdString(cmd.clientid()).simplified(), "REGISTER_ACCOUNT",
-                                         "Too many usernames registered with this email address", false);
+    if ((sqlInterface->checkNumberOfUserAccounts(emailAddress) >= servatrice->getMaxAccountsPerEmail())) {
```
Comment (ctrlaltca) [inline servatrice/src/serversocketinterface.cpp]: Is this change somewhat related/needed? It forces servatrice to make a query to lookup if an email address is already used even if the related settings is disabled.

File: cockatrice/src/remoteclient.h [hunk 3]
```diff
@@ -85,28 +87,35 @@ private slots:
...
 public:
-    RemoteClient(QObject *parent = 0);
-    ~RemoteClient();
+    explicit RemoteClient(QObject *parent = nullptr);
+    ~RemoteClient() override;
     QString peerName() const
     {
-        return socket->peerName();
+        if (usingWebSocket) {
+            return socket->peerName();
```
Comment (ctrlaltca) [inline cockatrice/src/remoteclient.h]: bug: inverted cases. If `usingWebSocket` you should return `websocket->peerName()`

File: cockatrice/src/remoteclient.cpp (hunk 1 above)
Comment (basicer) [reply cockatrice/src/remoteclient.cpp]: Was thinking the same thing.

File: cockatrice/src/remoteclient.h (hunk 3 above)
Comment (basicer) [reply cockatrice/src/remoteclient.h]: Good catch!

File: servatrice/src/serversocketinterface.cpp (hunk 2 above)
Comment (ZeldaZach) [reply servatrice/src/serversocketinterface.cpp]: Fixed
